### PR TITLE
[FEATURE] Allow subtracting from counting vector

### PR DESCRIPTION
**Pull request by marehr**
Hello seqan3-team,

I need an (efficient) way to remove `minimiser` (= `binning_bitvector` = `ibf.bulk_contains(minimiser)`) from a `counting_vector` again.

Furthermore, I added for symmetry the overload `counting_vector -= counting_vector`, too, but strictly speaking, I don't need it.

This PR adds the necessary overloads.

### CODE REVIEW

**Review by remyschwab — APPROVED** [review]

**Comment by eseiler** [comment]
Two things I'm wondering about:

### enumerate_bins

`enumerate_bins` would lead me to expect that I get something like a counted range. `enumerate` in python generates pairs of a number and an element of the range you are iterating over.

Something akin to `for_each_set_bin(binning_bitvector, callable)` seems more fitting?

### underflow

With the subtraction, you could underflow a counting_vector. I'd say, you usually want and use an unsigned integer for the counting_vector.
Would it make sense to add some checks for debug mode?

File: include/seqan3/search/dream_index/interleaved_bloom_filter.hpp [hunk 1]
```diff
@@ -817,55 +825,47 @@ class counting_vector : public std::vector<value_t>
...
-            for (size_t bin = bit_pos; bit_sequence != 0u; ++bin, bit_sequence >>= 1)
-            {
-                // Jump to the next 1 and
-                bin += jump_to_next_1bit(bit_sequence);
-
-                // increment the corresponding vector entry.
-                ++(*this)[bin];
-            }
-        }
+            --(*this)[bin];
```
Comment (marehr) [inline include/seqan3/search/dream_index/interleaved_bloom_filter.hpp:867]: > With the subtraction, you could underflow a counting_vector. I'd say, you usually want and use an unsigned integer for the counting_vector. Would it make sense to add some checks for debug mode?

```suggestion
            assert((*this)[bin] > 0);
            --(*this)[bin];
```

?

File: include/seqan3/search/dream_index/interleaved_bloom_filter.hpp [hunk 2]
```diff
@@ -887,6 +887,52 @@ class counting_vector : public std::vector<value_t>
...
+
+    /*!\brief Bin-wise substraction of two `seqan3::counting_vector`s.
+     * \param rhs The other seqan3::counting_vector.
+     * \attention The seqan3::counting_vector must be at least as big as `rhs`.
+     */
+    counting_vector & operator-=(counting_vector const & rhs)
+    {
+        assert(this->size() >= rhs.size()); // The counting vector may be bigger than what we need.
+
+        std::transform(this->begin(), this->end(), rhs.begin(), this->begin(), std::minus<value_t>());
```
Comment (marehr) [inline include/seqan3/search/dream_index/interleaved_bloom_filter.hpp:899]:
```suggestion
        std::transform(this->begin(), this->end(), rhs.begin(), this->begin(), [](auto a, auto b)
        {
            assert(a > b);
            return a - b;
        });
```

?

File: include/seqan3/search/dream_index/interleaved_bloom_filter.hpp [hunk 3]
```diff
@@ -817,55 +825,47 @@ class counting_vector : public std::vector<value_t>
...
+     * \param binning_bitvector The seqan3::interleaved_bloom_filter::membership_agent_type::binning_bitvector.
+     * \attention The counting_vector must be at least as big as `binning_bitvector`.
+     */
+    template <typename binning_bitvector_t>
+    //!\cond
+        requires is_binning_bitvector<binning_bitvector_t>
+    //!\endcond
+    counting_vector & operator-=(binning_bitvector_t const & binning_bitvector)
+    {
+        enumerate_bins(binning_bitvector, [this](size_t const bin)
```
Comment (marehr) [inline include/seqan3/search/dream_index/interleaved_bloom_filter.hpp:865]:
```suggestion
        for_each_set_bin(binning_bitvector, [this](size_t const bin)
```

File: include/seqan3/search/dream_index/interleaved_bloom_filter.hpp [hunk 4]
```diff
@@ -817,55 +825,47 @@ class counting_vector : public std::vector<value_t>
...
+        requires is_binning_bitvector<binning_bitvector_t>
     //!\endcond
-    counting_vector & operator+=(rhs_t const & rhs)
+    counting_vector & operator+=(binning_bitvector_t const & binning_bitvector)
     {
-        assert(this->size() >= rhs.size()); // The counting vector may be bigger than what we need.
-
-        // Jump to the next 1 and return the number of places jumped in the bit_sequence
-        auto jump_to_next_1bit = [] (size_t & x)
+        enumerate_bins(binning_bitvector, [this](size_t const bin)
```
Comment (marehr) [inline include/seqan3/search/dream_index/interleaved_bloom_filter.hpp:845]:
```suggestion
        for_each_set_bin(binning_bitvector, [this](size_t const bin)
```

File: include/seqan3/search/dream_index/interleaved_bloom_filter.hpp [hunk 5]
```diff
@@ -887,6 +887,52 @@ class counting_vector : public std::vector<value_t>
...
+        std::transform(this->begin(), this->end(), rhs.begin(), this->begin(), std::minus<value_t>());
+
+        return *this;
+    }
+
+private:
+
+    //!\brief Enumerates all bins of a seqan3::interleaved_bloom_filter::membership_agent_type::binning_bitvector.
+    template <typename binning_bitvector_t, typename on_bin_fn_t>
+    void enumerate_bins(binning_bitvector_t && binning_bitvector, on_bin_fn_t && on_bin_fn)
```
Comment (marehr) [inline include/seqan3/search/dream_index/interleaved_bloom_filter.hpp:908]:
```suggestion
    void for_each_set_bin(binning_bitvector_t && binning_bitvector, on_bin_fn_t && on_bin_fn)
```

**Review by eseiler — APPROVED** [review]

**Pull request merged**